{% extends "base.html" %}

{% block content %}
    <div id="page-wrapper">
        <div class="row">
            <div class="col-lg-12">
                <h1 class="page-header">sRNAtoolbox: a collection of small RNA analysis tools</h1>
            </div>
            <!-- /.col-lg-12 -->
        </div>
        <!-- /.row -->
        <div class="row">
            <div class="col-lg-12">
                <div class="well well-lg col-lg-12">
                    <h4>sRNAtoolbox is aimed to provide small RNA researchers with several useful tools including sRNA
                        expression profiling from deep sequencing experiments and several downstream analysis tools.
                        The center piece of sRNAtoolbox is sRNAbench, which allows the expression profiling and
                        prediction of novel microRNAs in deep sequencing experiments.
                        The other tools can be either launched on sRNAbench results, or independently using the
                        appropriate file formats.</h4><br>
                    <h4>Currently the toolbox is comprised of the following methods:</h4>


                </div>


                <div class="col-lg-8">

                    <div class="panel panel-green">
                        <div class="panel-heading">
                            <a href="/srnatoolbox/srnabench/" style="color: #ffffff"> <h3>sRNAbench:</h3> <h5>Expression profiling of small RNAs and prediction of novel microRNAs from deep sequencing data</h5>  </a>
                        </div>
                        <div class="panel-body">
                            <p>sRNAbench is principally designed for the use in model species, or at least species with
                                an available genome assembly or annotations is fasta format. It accepts several commonly
                                used input formats like fastq, sra, fasta or read/count. It performs the expression
                                profiling of known microRNAs and other types of small RNAs, detects and analysis
                                isomiRs, prepares several graphical summaries and predicts novel microRNAs both in
                                plants and animals.</p>
                        </div>
                        <div class="panel-footer">
                            <a href="/srnatoolbox/srnabench/"> Go to sRNAbench </a>
                        </div>
                    </div>
                </div>
                <div class="col-lg-8">
                    <div class="panel panel-yellow">
                        <div class="panel-heading">
                            <a href="/srnatoolbox/srnade/" style="color: #ffffff"> <h3>sRNAde:</h3> <h5>Differential expression analysis </h5></a>
                        </div>
                        <div class="panel-body">
                            <p>sRNAde is a module for the detection of differentially expressed sRNA. The input is
                                either a number of sRNAbench output folders or a user-given expression matrix.
                                For the detection of differentially expressed RNAs it applies 3 widely used methods,
                                edgeR, DEseq and noiseq. Apart of the output from the individual methods, it provides a
                                consensus differential expression file. Additionally, this tool can perform cluster
                                analysis (heatmaps), isomiR analysis and it gives a summary on the sequencing statistic
                                of all used samples (if the input has been from sRNAbench).</p>
                        </div>
                        <div class="panel-footer">
                            <a href="/srnatoolbox/srnade/"> Go to sRNAde </a>
                        </div>
                    </div>
                </div>
                <div class="col-lg-8">
                    <div class="panel panel-primary">
                        <div class="panel-heading">
                            <a href="/srnatoolbox/srnagfree/" style="color: #ffffff"> <h3>sRNAgFree:</h3>  <h5>Profiling and detection of novel microRNAs in non-model specie</h5>  </a>
                        </div>
                        <div class="panel-body">
                            <p>Currently, microRNA deep sequencing data is available for many species that do not count
                                with a proper genome assembly. In such cases, frequently simple homology based detection
                                methods are applied. These methods can lead to false positive predictions but will also
                                fail to detect microRNAs that are not present in the reference database (like miRBase).
                                sRNAgFree is aimed for the prediction and expression profiling of microRNAs in
                                non-model species. It based on the duplex properties formed by the guide and passenger
                                strand and applies homology and learning form known binding and Drosha/Dicer processing
                                patterns.</p>
                        </div>
                        <div class="panel-footer">
                            <a href="/srnatoolbox/srnagfree/"> Go to sRNAgFree </a>
                        </div>
                    </div>
                </div>
</div>
                <div class="row row-grid">
                <div class="col-lg-8">
                    <div class="panel panel-success">
                        <div class="panel-heading">
                            <a href="/srnatoolbox/srnablast/" style="color: #3c763d"> <h3>sRNAblast:</h3> <h5>Blast analysis of deep sequencing reads against a local nt/nr (NCBI link) database</h5>  </a>
                        </div>
                        <div class="panel-body">
                            <p>This tool is principally aimed for the analysis of reads that could not be mapped in
                                sRNAbench or other profiling tools. The results could either point towards contamination
                                sources or biological meaningful information like the presence of unexpected viral or
                                bacterial RNA molecules.</p>
                        </div>
                        <div class="panel-footer">
                            <a href="/srnatoolbox/srnablast/" style="color: #3c763d"> Go to sRNAblast </a>
                        </div>
                    </div>


                </div>

                <div class="col-lg-8">
                    <div class="panel panel-info">
                        <div class="panel-heading">
                            <a href="/srnatoolbox/mirconstarget/" style="color: #31708f"> <h3>miRNAconsTarget:</h3> <h5>predict microRNA targets on user defined sets of microRNAs and 3'UTRs </h5> </a>
                        </div>
                        <div class="panel-body">
                            <p>Most microRNA target prediction tools are implemented as webservers on a limited number of
                                species, or do allow only the prediction of a limited number of microRNAs and/or mRNAs.
                                miRNAconsTargets uses two microRNA target prediction programs, miranda and TargetSpy and
                                applies them to the user supplied sets of microRNAs and 3'UTRs. It reports both, the
                                individual predictions and a consensus prediction. Currently, only animal microRNA
                                targets can be predicted.</p>
                        </div>
                        <div class="panel-footer">
                            <a href="/srnatoolbox/mirconstarget/" style="color: #31708f"> Go to sRNAjBrowser </a>
                        </div>
                    </div>
                </div>

                <div class="col-lg-8">
                    <div class="panel panel-warning">
                        <div class="panel-heading">
                            <a href="/srnatoolbox/srnajbrowser/" style="color: #8a6d3b"> <h3>sRNAjBrowser:</h3> <h5>Browse the small RNA expression values in a genome context </h5> </a>
                        </div>
                        <div class="panel-body">
                            <p>In several small RNA research fields, the visual exploration of small RNA expression
                                pattern in a genome context will be helpful. sRNAjBrowser allows the user to analyse
                                the expression data in a genome context by means of a jBrowser implementation.
                                Furthermore, this tool is connected to NGSmethDB and can therefore also display
                                some selected methylation tracks</p>
                        </div>
                        <div class="panel-footer">
                            <a href="/srnatoolbox/srnajbrowser/" style="color: #8a6d3b"> Go to sRNAjBrowser </a>
                        </div>
                    </div>
                </div>





                    </div>

                <div class="row row-grid">
                <div class="col-lg-8">

                    <div class="panel panel-red">
                        <div class="panel-heading">
                            <a href="/srnatoolbox/srnafuncterms/" style="color: #ffffff"> <h3>sRNAfuncTerms:</h3> <h5>Determine over represented functional annotations in target gene set</h5>  </a>
                        </div>
                        <div class="panel-body">
                            <p>Overrepresented functional annotations in a list of target genes might give useful hints
                                on the functional implications of the microRNAs. sRNAfuncTerms takes a set of microRNAs
                                as input, retrieves the target genes from the underlying database and detected
                                overrepresented GO-terms among the target genes. Apart from the complete set of microRNAs,
                                it tests also all microRNA modules (combinations) which allow the user to detect the microRNAs
                                that act on the same pathways or share the same functions.</p>
                        </div>
                        <div class="panel-footer">
                            <a href="/srnatoolbox/srnafuncterms/"> Go to sRNAbench </a>
                        </div>
                    </div>
                </div>
                    </div>


		</div>



            </div>

    <script type="text/javascript">
    function showValue(newValue)
    {
        document.getElementById("range").innerHTML=newValue;
    }
    </script>

    <!-- /.row -->
        </div>
        <!-- /#page-wrapper -->
{% endblock %}


