{% extends "base.html" %}
{% load staticfiles %}
{% block content %}

    <div id="page-wrapper">
        <div class="row">
            <div class="col-lg-12">
                <div class="page-header">
                    {#                    <h1>Helper Tools: </h1>#}
                    <h1>{{tool}} </h1>
                </div>
            </div>
            <!-- /.col-lg-12 -->
        </div>
        <!-- /.row -->

        <div class="row">
            <div class="col-lg-9">



                {% if tool_url == "ensemlparser" %}
                    <div class="well">
                        <h4>Ensembl Parser:</h4>
                        <p>sRNAbench can read specially prepared annotations, i.e. the transcript name and the
                            classification separated by ‘:’. The 'Ensembl Parser' reads a Ensembl fasta annotation file
                            and generates the sRNAbench format. In can use cDNA and ncRNA fasta files from this
                        <a href="http://www.ensembl.org/info/data/ftp/index.html"> page </a>.


                        </p>

                        <p><a href="ftp://ftp.ensembl.org/pub/release-79/fasta/homo_sapiens/ncrna/Homo_sapiens.GRCh38.ncrna.fa.gz">Link to test data</a></p>
                       <p> If you use this data please cite: <a target="_blank" href="http://nar.oxfordjournals.org/content/41/D1/D48" >Paul Flicek et al. (2012). Ensembl 2013. Nucleic Acids Res.</a></p>

                    </div>

                {% endif %}


                {% if tool_url == "ncbiparser" %}
                    <div class="well">
                        <h4>NCBI Parser:</h4>
                        <p>sRNAbench can read specially prepared annotations, i.e. the transcript name and the
                            classification separated by ‘:’. The 'NCBI Parser' reads a NCBI fasta annotation file
                            and generates the sRNAbench format. In can use *.rna.fna.gz files from the following
                        <a href="ftp://ftp.ncbi.nlm.nih.gov/refseq/"> page </a>.


                        </p>

                        <p><a href="ftp://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/mRNA_Prot/human.1.rna.fna.gz">Link to test data</a></p>
                        <p> If you use this data please cite: <a target="_blank" href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3245008/" >Kim D. Pruitt et al. (2012). NCBI Reference Sequences (RefSeq): current status, new features and genome annotation policy. Nucleic Acids Res. PMCID: PMC3245008.</a></p>
                    </div>

                {% endif %}


                {% if tool_url == "rnacentralparser" %}
                    <div class="well">
                        <h4>RNA central parser:</h4>
                        <p>This parser extracts the non-coding sequences of a given species or taxonomy
                            level from the RNA central database version 2 (February 2015)
                            (<a href="http://rnacentral.org/">http://rnacentral.org/ </a>)
                            and prepares the sRNAbench format libraries.


                        </p>

                        <p>Test with: Hordeum vulgare</p>
                        <p> If you use this data please cite: <a target="_blank" href="http://nar.oxfordjournals.org/content/early/2014/10/28/nar.gku991.full" >The RNAcentral Consortium (2014) RNAcentral: an international database of ncRNA sequences. doi: 10.1093/nar/gku991</a></p>

                    </div>

                {% endif %}


                {% if tool_url == "trnaparser" %}
                    <div class="well">
                        <h4>genomic tRNA parser:</h4>
                        <p>tRNA annotations can be obtained parsing a species out of the genomic tRNA database
                            (<a href="http://gtrnadb.ucsc.edu/">http://gtrnadb.ucsc.edu/ </a>).


                        </p>

                        <p>Test with: Homo sapiens</p>
                        <p> If you use this data please cite: <a target="_blank" href="http://nar.oxfordjournals.org/content/37/suppl_1/D93.abstract" >Chan, P.P., Lowe, T.M. (2009) GtRNAdb: A database of transfer RNA genes detected in genomic sequence Nucl. Acids Res. 37(Database issue):D93-D97.</a></p>
                    </div>

                {% endif %}


                {% if tool_url == "removedup" %}
                    <div class="well">
                        <h4>Remove Duplicates from a Fasta File :</h4>
                        <p>The UCSC table browser allows to obtain 3' UTR sequences
                            which are needed when searching for microRNA target genes. However the output files have the following format:
                        </p>

                        <pre style="white-space: pre-line">
                            >hg19_refGene_NM_001184906 range=chr17:37408897-37417712 5'pad=0 3'pad=0 strand=- repeatMasking=none
                            CAATGGAGGTGGTCAACCTTGGCGAACTGAGTATTTAATGACACTTCTAG
                            AGCTACCGTGGAGTCTCTCCAGTGGAAGCAACCCCAGTGTTCTGAGCAAG
                        </pre>

                        <p>The name of the sequence ' hg19_refGene_NM_001184906' would not be recognized by downstream
                            analysis programs (functional enrichment analysis). This parser allows to substitute certain
                            parts of the sequence name in order to recover the name of the transcript ( NM_001184906 in this case). </p>

                        <p>Specifying 'hg19_refGene_' in the textbox would make the parser to remove this string
                            from the sequence name leaving only the name of the transcript.
                            This parser removes also duplicated IDs and gives the additional possibility to remove duplicated sequence.

                        </p>
                    </div>

                {% endif %}



                {% if tool_url == "extract" %}
                    <div class="well">
                        <h4>Extract Sequences from a fasta file:</h4>
                        <p>This parser salows to specify a search criterion for fasta sequence names. For example, microRNA sequences from miRBase start with the species name like 'has' (Homo sapiens). Providing
                            a miRBase fasta file specifying 'has' would make the program to return only human (hsa) sequences from the file. </p>
                    </div>

                {% endif %}





                    <!-- INPUT DATA -->
                    <div class="panel panel-default">
                        <div class="panel-heading">
                            <h4 class="panel-title">
                                <a data-toggle="collapse"  href="#collapseOne">Input data</a>
                            </h4>
                        </div>
                        <div id="collapseOne" class="panel-collapse collapse in">
                            <div class="panel-body">
                            <form action="/helper/run/{{tool_url}}" method="post" enctype="multipart/form-data">
                    {% csrf_token %}

                                {% if input %}
                                    <div class="form-group">
                                        <label> Upload input file({{input}}):</label>
                                        <input type="file" name="ifile" id=file1>
                                    </div>

                                    <div class="form-group">
                                        <label> Or provide a URL for big files </label>
                                        <label class="text-success"> (recommended!) </label>
                                        <input name="url" size="50">
                                    </div>

                                {% endif %}

                                {% if species %}

                                    <div class="form-group">
                                        <label> Provide a Species Name </label>
                                        <label class="text-success"> (Must be Scientific Name):</label>
                                        <input name="species" size="50">
                                    </div>
                                {% endif %}

                                {% if tax %}
                                    <div class="form-group">
                                        <label> Or provide a Taxonomy Name: </label>
                                        <input name="taxonomy" size="50">
                                    </div>
                                {% endif %}

                                {% if extract %}
                                    <div class="form-group">
                                        <label> Provide a string the characters used to select the sequences:</label>
                                        <input name="string" size="50">
                                    </div>
                                {% endif %}

                                {% if duplicates %}
                                    <div class="form-group">
                                        <label> Provide a string of characters to be drop out from the sequence names: </label>
                                        <input name="string" size="50">
                                    </div>


                                    <div class="form-group">
                                        <label><input type="checkbox" name="duplicates">Remove also duplicate </label>
                                        <label class="text-primary">SEQUENCES</label>
                                    </div>
                                {% endif %}

                            <input type="submit"  class="btn btn-primary" value="Submit">
                            <input type="reset"  class="btn btn-default" value="Reset">
                             </form>
                            </div>
                        </div>
                    </div>














            </div>


            <div class="col-lg-3">
                <div class="panel panel-info">
                    <div class="panel-heading">
                        <h4 class="panel-title">
                            Help
                        </h4>
                    </div>
                    <div class="panel-body">
                        <a href="{{ settings.MANUAL_URL }}" target="_blank">Link to Web Manual</a>

                    </div>
                </div>

            </div>


            <!-- /#page-wrapper -->
        </div>
    </div>
    </div>

{% endblock %}

{% block js %}
    <script>
        // tooltip demo
        $('.row').tooltip({
            selector: "[data-toggle=tooltip]",
            container: "body"
        })

    </script>

{% endblock %}
